src/queue/descriptor_chain.rs

```diff
@@ -178,3 +185,838 @@ impl<M: GuestAddressSpace> Iterator for DescriptorChainRwIter<M> {
+        let mut bufs = Vec::with_capacity(self.buffers.len());
+        for &vs in &self.buffers {
+            if buflen >= count {
+                break;
+            }
+
+            bufs.push(vs);
+
+            let rem = count - buflen;
+            if rem < vs.len() {
+                buflen += rem;
+            } else {
+                buflen += vs.len() as usize;
+            }
+        }
+
+        if bufs.is_empty() {
+            return Ok(0);
+        }
+
+        let bytes_consumed = f(&*bufs)?;
+
+        // This can happen if a driver tricks a device into reading/writing more data than
+        // fits in a `usize`.
+        let total_bytes_consumed =
+            self.bytes_consumed
+                .checked_add(bytes_consumed)
+                .ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::InvalidData, Error::DescriptorChainOverflow)
+                })?;
+
+        let mut rem = bytes_consumed;
+        while let Some(vs) = self.buffers.pop_front() {
+            if rem < vs.len() {
+                // Split the slice and push the remainder back into the buffer list. Safe because we
+                // know that `rem` is not out of bounds due to the check and we checked the bounds
+                // on `vs` when we added it to the buffer list.
+                self.buffers.push_front(vs.offset(rem).unwrap());
+                break;
+            }
+
+            // No need for checked math because we know that `vs.size() <= rem`.
+            rem -= vs.len();
+        }
+
+        self.bytes_consumed = total_bytes_consumed;
+
+        Ok(bytes_consumed)
+    }
+
+    fn split_at(&mut self, offset: usize) -> Result<DescriptorChainConsumer<'a>> {
+        let mut rem = offset;
+        let pos = self.buffers.iter().position(|vs| {
+            if rem < vs.len() {
+                true
+            } else {
+                rem -= vs.len();
+                false
+            }
+        });
+
+        if let Some(at) = pos {
+            let mut other = self.buffers.split_off(at);
+
+            if rem > 0 {
+                // There must be at least one element in `other` because we checked
+                // its `size` value in the call to `position` above.
+                let front = other.pop_front().expect("empty VecDeque after split");
+                let (start, end) = front.split_at(rem).map_err(Error::VolatileMemoryError)?;
+                self.buffers.push_back(start);
+                other.push_front(end);
+            }
+
+            Ok(DescriptorChainConsumer {
+                buffers: other,
+                bytes_consumed: 0,
+            })
+        } else if rem == 0 {
+            Ok(DescriptorChainConsumer {
+                buffers: VecDeque::new(),
+                bytes_consumed: 0,
+            })
+        } else {
+            Err(Error::DescriptorChainSplitOOB(offset))
+        }
+    }
+}
+
+fn collect_desc_chain_buffers<M: GuestAddressSpace>(
+    mem: &GuestMemoryMmap,
```

Comment:
The guest memory may not be necessarily the same as the descriptor chain memory.

Reply:
Can you please provide some more details about why this ends up being the case? Having what are effectively two different handles to memory objects here (or in similar situations) is quite confusing, and it may be an indication that we need to adjust some of the abstractions we're defining in `vm-virtio`.

Reply:
@alexandruag  Ok, I tryed to get it to work with the memory from the DescriptorChain. My conclusion is that it's really hard, given that the GuestAddressSpace lifetime is associated with it. Even after significant effort to I didn't manage to refactor the code and use appropriate lifetime. collect_desc_chain_buffers() cannot infer a lifetime or take away the underlying GuestAddressSpace. Quite difficult tbh, and less powerful overall than having a seperate mem argument. Could you give it a try? If you manage to convince yourself it isn't possible as well, we may just leave it that way.

Reply:
Sure, I'll give it a shot as well.